src/Composer/Util/Http/ProxyHandler.php

```diff
@@ -0,0 +1,212 @@
+<?php declare(strict_types=1);
+
+/*
+ * This file is part of Composer.
+ *
+ * (c) Nils Adermann <naderman@naderman.de>
+ *     Jordi Boggiano <j.boggiano@seld.be>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Composer\Util\Http;
+
+use Composer\Downloader\TransportException;
+use Composer\Util\NoProxyPattern;
+
+/**
+ * @internal
+ * @author John Stevenson <john-stevenson@blueyonder.co.uk>
+ */
+class ProxyHandler
+{
+    /** @var ?string */
+    private $error = null;
+    /** @var ?ProxyItem */
+    private $httpProxy = null;
+    /** @var ?ProxyItem */
+    private $httpsProxy = null;
+    /** @var ?NoProxyPattern */
+    private $noProxyHandler = null;
+
+    /** @var ?ProxyHandler */
+    private static $instance = null;
+
+    /** The following 3 properties can be removed after the transition period */
+
+    /** @var bool */
+    private $ignoreHttpsProxy = false;
+    /** @var bool */
+    private $isTransitional = false;
+    /** @var bool */
+    private $needsTransitionWarning = false;
+
+    private function __construct()
+    {
+        // this can be removed after the transition period
+        $this->isTransitional = true;
+
+        try {
+            $this->getProxyData();
+        } catch (\RuntimeException $e) {
+            $this->error = $e->getMessage();
+        }
+    }
+
+    public static function getInstance(): ProxyHandler
+    {
+        if (self::$instance === null) {
+            self::$instance = new self();
+        }
+
+        return self::$instance;
+    }
+
+    /**
+     * Clears the persistent instance
+     */
+    public static function reset(): void
+    {
+        self::$instance = null;
+    }
+
+    /**
+     * Returns a RequestProxy instance for the request url
+     *
+     * @param non-empty-string $requestUrl
+     */
+    public function getProxyForRequest(string $requestUrl): RequestProxy
+    {
+        if ($this->error !== null) {
+            throw new TransportException('Unable to use a proxy: '.$this->error);
+        }
+
+        $scheme = (string) parse_url($requestUrl, PHP_URL_SCHEME);
+        $url = null;
+        $auth = null;
+        $options = null;
+        $info = null;
+
+        $proxy = $this->getProxyForScheme($scheme);
+
+        if ($proxy !== null) {
+            if ($this->noProxy($requestUrl)) {
+                $info = 'excluded by no_proxy';
+            } else {
+                $url = $proxy->getProxyUrl();
+                $auth = $proxy->getCurlAuth();
+                $options = $proxy->getContextOptions($scheme);
+                $info = $proxy->getSafeUrl();
+            }
+        }
+
+        return new RequestProxy($url, $auth, $options, $info);
```

Comment:
IMO this would read better with a few factory methods hiding all the copying around of stuff:

```suggestion
        $proxy = $this->getProxyForScheme($scheme);

        if ($proxy === null) {
            return RequestProxy::none();
        }

        if ($this->noProxy($requestUrl)) {
            return RequestProxy::noProxy();
        }

        return RequestProxy::fromProxyItem($scheme);
```

Reply:
Neat. I really like this and will implement.